src/Prefetch.cpp

```diff
@@ -0,0 +1,228 @@
+        Expr prefetch_load = make_similar_load(body, buf_name, indices);
+        internal_assert(prefetch_load.defined());
+        Type type = prefetch_load.type();
+        Expr prefetch_addr = Call::make(Handle(), Call::address_of, {prefetch_load}, Call::Intrinsic);
+
+        Stmt prefetch;
+        Expr stride_0 = Variable::make(Int(32), buf_name + ".stride.0");
+        // TODO: This is only correct if stride_0 is 1. We need to assert that this is true.
+        Expr extent_0_bytes = prefetch_extent[0] * type.bytes();
+        if (box.size() == 1) {
+            // The prefetch is only 1 dimensional, just emit a flat prefetch.
+            prefetch = Evaluate::make(Call::make(Int(32), Call::prefetch,
+                                                 {prefetch_addr, extent_0_bytes},
+                                                 Call::PureIntrinsic));
+        } else {
+            // Make a 2D prefetch.
+            Expr stride_1 = Variable::make(Int(32), buf_name + ".stride.1");
+            Expr stride_1_bytes = stride_1 * type.bytes();
+            prefetch = Evaluate::make(Call::make(Int(32), Call::prefetch_2d,
+                                                 {prefetch_addr, extent_0_bytes, prefetch_extent[1], stride_1_bytes},
+                                                 Call::PureIntrinsic));
+
+            // Make loops for the rest of the dimensions (possibly zero).
+            for (size_t i = 2; i < box.size(); i++) {
+                prefetch = For::make(index_names[i], prefetch_min[i], prefetch_extent[i],
+                                     ForType::Serial, DeviceAPI::None,
+                                     prefetch);
+            }
+        }
+
+        // We should only prefetch buffers that are used.
+        if (box.maybe_unused()) {
+            prefetch = IfThenElse::make(box.used, prefetch);
+        }
+
+        return Block::make({prefetch, body});
+    }
+
+    void visit(const For *op) {
+        // Add loop variable to interval scope for any inner loop prefetch
+        Expr loop_var = Variable::make(Int(32), op->name);
+        bounds.push(op->name, Interval(loop_var, loop_var));
+        Stmt body = mutate(op->body);
+        bounds.pop(op->name);
+
+        if (prefetches) {
+            for (const Prefetch &p : *prefetches) {
+                if (!ends_with(op->name, "." + p.var)) {
+                    continue;
+                }
+                // Add loop variable + prefetch offset to interval scope for box computation
+                Expr fetch_at = loop_var + p.offset;
+                bounds.push(op->name, Interval(fetch_at, fetch_at));
+                map<string, Box> boxes_read = boxes_required(body, bounds);
+                bounds.pop(op->name);
+
+                // Don't prefetch buffers that are written to. We assume that these already
+                // have good locality.
+                // TODO: This is not a good assumption. It would be better to have the
```

Comment:
I was thinking we could do this in a later PR, but we should probably do it now. It will change the prefetch scheduling directive interface.